test/ably/realtime/realtimeauth_test.py

```diff
@@ -0,0 +1,153 @@
+        await ably.close()
+
+    async def test_auth_with_auth_callback_token_with_details(self):
+        rest = await TestApp.get_ably_rest()
+
+        async def callback(params):
+            token_details = await rest.auth.request_token(token_params=params)
+            return token_details
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        response_time_ms = await ably.connection.ping()
+        assert response_time_ms is not None
+        assert ably.connection.error_reason is None
+        await ably.close()
+
+    async def test_auth_with_auth_callback_with_token_string(self):
+        rest = await TestApp.get_ably_rest()
+
+        async def callback(params):
+            token_details = await rest.auth.request_token(token_params=params)
+            return token_details.token
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        response_time_ms = await ably.connection.ping()
+        assert response_time_ms is not None
+        assert ably.connection.error_reason is None
+        await ably.close()
+
+    async def test_auth_with_auth_callback_invalid_token(self):
+        async def callback(params):
+            return "invalid token"
+
+        ably = await TestApp.get_ably_realtime(auth_callback=callback)
+        state_change = await ably.connection.once_async(ConnectionState.FAILED)
+        assert state_change.reason.code == 40005
+        assert state_change.reason.status_code == 400
+        await ably.close()
+
+    async def test_auth_with_auth_url_json(self):
+        echo_url = 'https://echo.ably.io'
+        rest = await TestApp.get_ably_rest()
+        token_details = await rest.auth.request_token()
+        token_details_json = json.dumps(token_details.to_dict())
+        url_path = f"{echo_url}/?type=json&body={token_details_json}"
+
+        ably = await TestApp.get_ably_realtime(auth_url=url_path)
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        response_time_ms = await ably.connection.ping()
+        assert response_time_ms is not None
+        assert ably.connection.error_reason is None
+        await ably.close()
+
+    async def test_auth_with_auth_url_text_plain(self):
+        echo_url = 'https://echo.ably.io'
+        rest = await TestApp.get_ably_rest()
+        token_details = await rest.auth.request_token()
+        url_path = f"{echo_url}/?type=text&body={token_details.token}"
+
+        ably = await TestApp.get_ably_realtime(auth_url=url_path)
+        await ably.connection.once_async(ConnectionState.CONNECTED)
+        response_time_ms = await ably.connection.ping()
+        assert response_time_ms is not None
+        assert ably.connection.error_reason is None
+        await ably.close()
+
+    async def test_auth_with_auth_url_post(self):
```

Comment:
Should be able to fix this by:
converting auth_params to a dict:
```python
if isinstance(auth_params, TokenDetails):
    auth_params = auth_params.to_dict()
```
and then catching capability parsing errors like so:
```python
    try:
        self.__capability = Capability(json.loads(capability))
    except json.JSONDecodeError:
        self.__capability = Capability(json.loads(capability.replace("'", '"')))
```